expose.go

```diff
@@ -0,0 +1,160 @@
+			return err
+		}))
+		if err != nil {
+			return
+		}
+
+		ListenTarget(ctx, func(ev interface{}) {
+			switch ev := ev.(type) {
+			case *runtime.EventBindingCalled:
+				var payload BindingCalledPayload
+
+				err := json.Unmarshal([]byte(ev.Payload), &payload)
+				if err != nil {
+					return
+				}
+
+				var expression string
+
+				c.Target.bindingFuncMu.RLock()
+				defer c.Target.bindingFuncMu.RUnlock()
+				if fn, ok := c.Target.bindingFuncs[payload.Name]; ok {
+					res, err := fn(payload.Args)
+					if err != nil {
+						expression = deliverError(payload.Name, payload.Seq, err.Error(), err.Error())
+					} else {
+						expression = deliverResult(payload.Name, payload.Seq, res)
+					}
+				} else {
+					expression = deliverError(payload.Name, payload.Seq, "bindingCall name not exsit", "")
+				}
+
+				go func() {
+					Run(ctx, Evaluate(expression, nil, func(p *runtime.EvaluateParams) *runtime.EvaluateParams {
+						return p.WithContextID(ev.ExecutionContextID)
+					}))
+				}()
+
+			}
+		})
+	})
+
+	c.Target.bindingFuncMu.Lock()
+	if _, ok := c.Target.bindingFuncs[fnName]; ok {
+		c.Target.bindingFuncMu.Unlock()
+		return ErrExposeNameExist
+	}
+	c.Target.bindingFuncs[fnName] = fn
+	c.Target.bindingFuncMu.Unlock()
+
+	err := Run(ctx, runtime.AddBinding(fnName))
+	if err != nil {
+		return err
+	}
+
+	expression := addPageBinding("exposedFun", fnName)
+	err = Run(ctx, ActionFunc(func(ctx context.Context) error {
+		_, err := page.AddScriptToEvaluateOnNewDocument(expression).Do(ctx)
+		return err
+	}))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+const exposedFunJS = `
+function deliverError(name, seq, message, stack) {
+	const error = new Error(message);
+	error.stack = stack;
+	window[name].callbacks.get(seq).reject(error);
+	window[name].callbacks.delete(seq);
+}
+
+function deliverResult(name, seq, result) {
+	window[name].callbacks.get(seq).resolve(result);
+	window[name].callbacks.delete(seq);
+}
+
+function addPageBinding(type, name) {
+	// This is the CDP binding.
+	const callCDP = self[name];
+	console.log("callCDP",callCDP)
+	// We replace the CDP binding with a Puppeteer binding.
+	Object.assign(self, {
+		[name](args) {
+			if(typeof args != "string"){
+				return Promise.reject(new Error('function takes exactly one argument, this argument should be string'))
+			}
+			var _a, _b;
+			// This is the Puppeteer binding.
+			const callPuppeteer = self[name];
+			(_a = callPuppeteer.callbacks) !== null && _a !== void 0 ? _a : (callPuppeteer.callbacks = new Map());
+			const seq = ((_b = callPuppeteer.lastSeq) !== null && _b !== void 0 ? _b : 0) + 1;
+			callPuppeteer.lastSeq = seq;
+			callCDP(JSON.stringify({ type, name, seq, args }));
+			return new Promise((resolve, reject) => {
+				callPuppeteer.callbacks.set(seq, { resolve, reject });
+			});
+		},
+	});
+}
+`
+
+func deliverError(name string, seq int64, message, stack string) string {
+	var cmd string = `deliverError("%s",%d,"%s","%s");`
```

Comment:
I think it's not safe to construct js code like this. Maybe `chromedp.CallFunctionOn` can be used here?

Reply:
ok. I didn't know this function existed before.

Reply:
it has been modified to CallFunctionOn. 
By the way, I'd like to know how to use chromedp.CallFunctionOn without executionContextId or objectId?

Reply:
According to [Runtime.callFunctionOn](https://chromedevtools.github.io/devtools-protocol/tot/Runtime/#method-callFunctionOn):

> Either objectId or executionContextId should be specified.